File: ts/src/actions/FlowUpdate.ts
import type {ActionFlowUpdate} from "@code0-tech/tucana/aquila";
import {CodeZeroEvent} from "../events";
import type {Action} from "../action";

export const packetType = "flowUpdate";

export function handle(action: Action, update: ActionFlowUpdate): void {
    if (update.data.oneofKind === "updatedFlow") {
        const flow = update.data.updatedFlow;
        action.flows.set(flow.flowId, flow);
        action.emit(CodeZeroEvent.flowUpdated, flow);
    } else if (update.data.oneofKind === "deletedFlow") {
        const flowId = update.data.deletedFlow;
        action.flows.delete(flowId);
        action.emit(CodeZeroEvent.flowDeleted, flowId);
    }
}

File: py/hercules/actions/flow_update.py
"""Flow update handler (port of ``src/actions/FlowUpdate.ts``)."""
from __future__ import annotations

from hercules.events import CodeZeroEvent

packet_type = "flow_update"


def handle(action, update) -> None:
    which = update.WhichOneof("data")
    if which == "updated_flow":
        flow = update.updated_flow
        action.flows.set(flow.flow_id, flow)
        action.emit(CodeZeroEvent.flow_updated, flow)
    elif which == "deleted_flow":
        flow_id = update.deleted_flow
        action.flows.delete(flow_id)
        action.emit(CodeZeroEvent.flow_deleted, flow_id)
